### PR TITLE
Added FIRK solvers

### PR DESCRIPTION
**Pull request by ParamThakkar123**
## Checklist

- [X] Appropriate tests were added
- [X] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [X] Any new documentation only uses public API
  
## Additional context

solves a part of #2177

### CODE REVIEW

File: src/OrdinaryDiffEq.jl [hunk 1]
```diff
@@ -149,14 +149,12 @@ include("generic_rosenbrock.jl")
...
 include("caches/high_order_rk_caches.jl")
-include("caches/sdirk_caches.jl")
-include("caches/firk_caches.jl")
-include("caches/kencarp_kvaerno_caches.jl")
 include("caches/linear_caches.jl")
 include("caches/linear_nonlinear_caches.jl")
 include("caches/rosenbrock_caches.jl")
 include("caches/adams_bashforth_moulton_caches.jl")
 include("caches/nordsieck_caches.jl")
+include("../lib/OrdinaryDiffEqSDIRK/src/sdirk_caches.jl")
```
Comment (ChrisRackauckas) [inline src/OrdinaryDiffEq.jl:157]:
```suggestion
```

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 2]
```diff
@@ -0,0 +1,24 @@
+module OrdinaryDiffEqFIRK
+
+import OrdinaryDiffEq: alg_order, calculate_residuals!,
+                       initialize!, perform_step!, @unpack, unwrap_alg,
+                       calculate_residuals,
+                       OrdinaryDiffEqAlgorithm,
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:6]:
```suggestion
                       OrdinaryDiffEqAlgorithm, OrdinaryDiffEqNewtonAdaptiveAlgorithm,
```

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import OrdinaryDiffEq: alg_order, calculate_residuals!,
+                       initialize!, perform_step!, @unpack, unwrap_alg,
+                       calculate_residuals,
+                       OrdinaryDiffEqAlgorithm, OrdinaryDiffEqNewtonAdaptiveAlgorithm,
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new,
+                       explicit_rk_docstring, trivial_limiter!,
+                       _ode_interpolant!, _ode_addsteps!
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:12]:
```suggestion
                       _ode_interpolant!, _ode_addsteps!, AbstractController
```

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+                       initialize!, perform_step!, @unpack, unwrap_alg,
+                       calculate_residuals,
+                       OrdinaryDiffEqAlgorithm, OrdinaryDiffEqNewtonAdaptiveAlgorithm,
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new,
+                       explicit_rk_docstring, trivial_limiter!,
+                       _ode_interpolant!, _ode_addsteps!, AbstractController
+
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:13]:
```suggestion
                       _ode_interpolant!, _ode_addsteps!, AbstractController,
                       NLStatus

```

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+                       initialize!, perform_step!, @unpack, unwrap_alg,
+                       calculate_residuals,
+                       OrdinaryDiffEqAlgorithm, OrdinaryDiffEqNewtonAdaptiveAlgorithm,
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new,
+                       explicit_rk_docstring, trivial_limiter!,
+                       _ode_interpolant!, _ode_addsteps!, AbstractController,
+                       NLStatus
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:13]:
```suggestion
                       _ode_interpolant!, _ode_addsteps!, AbstractController,
                       NLStatus, qmax_default, alg_adaptive_order
```

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+                       calculate_residuals,
+                       OrdinaryDiffEqAlgorithm, OrdinaryDiffEqNewtonAdaptiveAlgorithm,
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new,
+                       explicit_rk_docstring, trivial_limiter!,
+                       _ode_interpolant!, _ode_addsteps!, AbstractController,
+                       NLStatus, qmax_default, alg_adaptive_order
+
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:14]:
```suggestion
import FastBroadcast: @..
```

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas In this PR all the UnDef errors have been solved. This now has some precision and method call errors in some files

**Comment by ParamThakkar123** [comment]
I have also started a PR for DAE solvers #2292

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import FastBroadcast: @..
+include("algorithms.jl")
+include("alg_utils.jl")
+include("controllers.jl")
+include("firk_caches.jl")
+include("firk_tableaus.jl")
+include("firk_perform_step.jl")
+include("integrator_interface.jl")
+
+export RadauIIA3, RadauIIA5
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:30]:
```suggestion
export RadauIIA3, RadauIIA5, RadauIIA7
```

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new,
+                       explicit_rk_docstring, trivial_limiter!,
+                       _ode_interpolant!, _ode_addsteps!, AbstractController,
+                       NLStatus, qmax_default, alg_adaptive_order, DEFAULT_PRECS,
+                       UJacobianWrapper, build_J_W, build_jac_config, UDerivativeWrapper,
+                       Convergence, calc_J!, dolinsolve, FastConvergence, calc_J,
+                       mul!
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:16]:
```suggestion
                       mul!, stepsize_controller!
```

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new,
+                       explicit_rk_docstring, trivial_limiter!,
+                       _ode_interpolant!, _ode_addsteps!, AbstractController,
+                       NLStatus, qmax_default, alg_adaptive_order, DEFAULT_PRECS,
+                       UJacobianWrapper, build_J_W, build_jac_config, UDerivativeWrapper,
+                       Convergence, calc_J!, dolinsolve, FastConvergence, calc_J,
+                       mul!, stepsize_controller!
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:16]:
```suggestion
                       stepsize_controller!, islinearfunction
```

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+using MuladdMacro, DiffEqBase, RecursiveArrayTools
+using SciMLOperators: AbstractSciMLOperator
+using LinearAlgebra: I, UniformScaling
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:19]:
```suggestion
using LinearAlgebra: I, UniformScaling, mul!, lu
```

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new,
+                       explicit_rk_docstring, trivial_limiter!,
+                       _ode_interpolant!, _ode_addsteps!, AbstractController,
+                       NLStatus, qmax_default, alg_adaptive_order, DEFAULT_PRECS,
+                       UJacobianWrapper, build_J_W, build_jac_config, UDerivativeWrapper,
+                       Convergence, calc_J!, dolinsolve, FastConvergence, calc_J,
+                       stepsize_controller!, islinearfunction
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:16]:
```suggestion
                       stepsize_controller!, islinearfunction, step_accept_controller!, step_reject_controller!
```

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas any changes or fixes needed on this PR ??

**Comment by ChrisRackauckas** [comment]
Tests don't pass. The PredictiveController should not have been moved. Move that back, and then the spot that does `if radau` should be `if isfirk(alg)`, and then in the alg_utils you should define `isfirk(alg) = false`, and then in the OrdinaryDiffEqFIRK/alg_utils.jl add `isfirk(alg:: radau algorithms) = true`.

File: lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+                       constvalue, _unwrap_val, du_alias_or_new,
+                       explicit_rk_docstring, trivial_limiter!,
+                       _ode_interpolant!, _ode_addsteps!, AbstractController,
+                       NLStatus, qmax_default, alg_adaptive_order, DEFAULT_PRECS,
+                       UJacobianWrapper, build_J_W, build_jac_config, UDerivativeWrapper,
+                       Convergence, calc_J!, dolinsolve, FastConvergence, calc_J,
+                       stepsize_controller!, islinearfunction, step_accept_controller!, step_reject_controller!,
+                       PredictiveController, alg_can_repeat_jac, NewtonAlgorithm, fac_default_gamma,
+                       get_new_W_γdt_cutoff, get_current_adaptive_order, VerySlowConvergence, 
+                       Divergence
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqFIRK/src/OrdinaryDiffEqFIRK.jl:19]:
```suggestion
                       Divergence, isfirk
```

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**